src/core/core.scale.js

```diff
@@ -311,39 +378,37 @@ module.exports = Element.extend({
 	},
 	calculateTickRotation: function() {
 		var me = this;
-		var context = me.ctx;
-		var tickOpts = me.options.ticks;
-		var labels = labelsFromTicks(me._ticks);
+		var options = me.options;
+		var tickOpts = options.ticks;
+		var ticks = me.getTicks();
+		var labelSizes, maxLabelWidth, maxLabelHeight, tickWidth, angleRadians, cosRotation, sinRotation;
 
 		// Get the width of each grid by calculating the difference
 		// between x offsets between 0 and 1.
-		var tickFont = helpers.options._parseFont(tickOpts);
-		context.font = tickFont.string;
-
 		var labelRotation = tickOpts.minRotation || 0;
 
-		if (labels.length && me.options.display && me.isHorizontal()) {
-			var originalLabelWidth = helpers.longestText(context, tickFont.string, labels, me.longestTextCache);
-			var labelWidth = originalLabelWidth;
-			var cosRotation, sinRotation;
+		if (ticks.length > 1 && options.display && me.isHorizontal()) {
+			labelSizes = computeLabelSizes(me.ctx, parseTickFontOptions(tickOpts), ticks, me.longestTextCache);
+			maxLabelWidth = labelSizes.widest.width;
+			maxLabelHeight = labelSizes.highest.height;
+
+			tickWidth = (me.chart.width - maxLabelWidth) / (ticks.length - 1);
 
 			// Allow 3 pixels x2 padding either side for label readability
-			var tickWidth = me.getPixelForTick(1) - me.getPixelForTick(0) - 6;
-
-			// Max label rotation can be set or default to 90 - also act as a loop counter
-			while (labelWidth > tickWidth && labelRotation < tickOpts.maxRotation) {
-				var angleRadians = helpers.toRadians(labelRotation);
-				cosRotation = Math.cos(angleRadians);
-				sinRotation = Math.sin(angleRadians);
-
-				if (sinRotation * originalLabelWidth > me.maxHeight) {
-					// go back one step
-					labelRotation--;
-					break;
+			if (maxLabelWidth > tickWidth - 6) {
```

Comment:
to avoid loop, this block could be replaced with something like: 
```js
var tickMarkLength = me.options.gridLines.tickMarkLength;
var maxHeight = me.maxHeight - tickMarkLength; // - scaleLabelSpace(me.options.scaleLabel)
var maxLabelDiagonal = Math.sqrt(maxLabelWidth * maxLabelWidth + maxLabelHeight * maxLabelHeight);
var size = calcRotatedBoxSize(maxLabelWidth , maxLabelHeight , helpers.toRadians(minRotation));
if (size.h < maxHeight && size.w > tickWidth) {
	var byWidth= Math.asin(Math.Min(maxLabelHeight / tickWidth, 1);
	var byHeight = Math.asin(Math.Min(maxLabelDiagonal / maxHeight, 1);
	labelRotation = helpers.toDegrees(Math.min(byWidth, byHeight));
}
labelRotation = Math.max(minRotation, Math.min(maxRotation, labelRotation));`
```
(`scaleLabelSpace`  was for issue #5906):

calcRotatedBoxSize:
```js
/**
 * Calculate rotated box size
 * @param {number} w - width of the box
 * @param {number} h - height of the box
 * @param {number} rad - rotation in radians
 * @return {object} - {w, h}
 */
function calcRotatedBoxSize(w, h, rad) {
	var cos = Math.cos(rad);
	var sin = Math.sin(rad);
	return {
		w: w * cos + h * sin,
		h: w * sin + h * cos
	};
}
```

Reply:
The loop was replaced with a logic similar to above.